src/sage/graphs/matching_covered_graph.py

```diff
@@ -0,0 +1,1228 @@
+          but the resulting graph after deletion of all of those is not
+          matching covered, then a :exc:`ValueError` exception is raised
+          without any alterations to the existing (matching covered) graph,
+          otherwise the vertices are deleted and nothing is returned.
+
+        EXAMPLES:
+
+        Providing with a list of vertices with at least one non-existent
+        vertex::
+
+            sage: S = graphs.StaircaseGraph(4)
+            sage: S
+            Staircase graph: Graph on 8 vertices
+            sage: G = MatchingCoveredGraph(S)
+            sage: G
+            Matching covered staircase graph: graph on 8 vertices
+            sage: T = list(range(5, 20, 2))
+            sage: G.delete_vertices(T)
+            Traceback (most recent call last):
+            ...
+            ValueError: vertex (9) not in the graph
+
+        Providing with a list of existent vertices whose deletion results in a
+        graph which is not matching covered::
+
+            sage: S = graphs.StaircaseGraph(4)
+            sage: S
+            Staircase graph: Graph on 8 vertices
+            sage: G = MatchingCoveredGraph(S)
+            sage: G
+            Matching covered staircase graph: graph on 8 vertices
+            sage: T = [1, 4]
+            sage: G.delete_vertices(T)
+            Traceback (most recent call last):
+            ...
+            ValueError: the resulting graph after the removal of the vertices is not matching covered
+
+        Providing with a list of existent vertices after the deletion of which
+        the resulting graph is still matching covered; note that in the
+        following example, after the deletion of two vertices from a staircase
+        graph, the resulting graph is NOT a staircase graph
+        (see :issue:`38768`)::
+
+            sage: S = graphs.StaircaseGraph(4)
+            sage: S
+            Staircase graph: Graph on 8 vertices
+            sage: G = MatchingCoveredGraph(S)
+            sage: G
+            Matching covered staircase graph: graph on 8 vertices
+            sage: T = [6, 7]
+            sage: G.delete_vertices(T)
+            sage: G  # Matching covered graph on 6 vertices
+            Matching covered staircase graph: graph on 6 vertices
+        """
+
+        for vertex in vertices:
+            if vertex not in self:
+                raise ValueError('vertex (%s) not in the graph' % str(vertex))
+
+        try:
+            G = Graph(self)
+            G.delete_vertices(vertices)
+            G_simple = G.to_simple()
+
+            M = Graph(self.get_matching())
+
+            if M:
+                M.delete_vertices(vertices)
+                # The resulting matching after the removal of the input vertices
+                # must be a valid perfect matching of the resulting graph obtained
+                # after the removal of the vertices
+
+                if any(d != 1 for d in G_simple.degree()) \
```

Comment:
I don't understand this test. For instance, how could an edge of `M` not be in `G_simple` ?

Reply:
this has not been answered yet.

Reply:
Yes. This is my bad. I have corrected it.